plugin/role_set_utils.go

```diff
@@ -0,0 +1,161 @@
+package gcpsecrets
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-gcp-common/gcputil"
+	"github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/useragent"
+	"github.com/hashicorp/vault/sdk/logical"
+	"google.golang.org/api/iam/v1"
+	"google.golang.org/api/option"
+	"regexp"
+	"time"
+)
+
+// roleSetResources is a wrapper around a roleset's managed GCP service account, IAM bindings, and/or keys.
+// It exists separately from roleset to allow for construction with or without a roleset (i.e. for deleting
+// accounts removed from a RoleSet), and because changing RoleSet requires migration of stored RoleSets.
+// A utility for creating roleSetResources from a RoleSet is provided below.
+type roleSetResources struct {
+	rolesetName string
+
+	accountId *gcputil.ServiceAccountId
+	bindings  ResourceBindings
+	tokenGen  *TokenGenerator
+}
+
+func (rs *RoleSet) boundResources() *roleSetResources {
+	if rs.AccountId == nil {
+		return nil
+	}
+	return &roleSetResources{
+		accountId:   rs.AccountId,
+		rolesetName: rs.Name,
+		bindings:    rs.Bindings,
+		tokenGen:    rs.TokenGen,
+	}
+}
+
+// addWalsForRoleSetResources creates WALs to clean up a roleset's service account, bindings, and a key if needed.
+func (b *backend) addWalsForRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) error {
+	if boundResources == nil {
```

Comment:
Should we behave the same if `boundResources` is empty? If you change this to a length check, it'll handle both.

Reply:
welp this is confusing - boundResources is `*rolesetResources`, a new object I created to wrap:

- a service account
- ResourceBindings
- TokenGen (key)

It's handier than passing around a whole roleset and allows me to remove some of the confusion where it wasn't clear which updated fields were updated in the RoleSet vs being passed into the function

Reply:
I am very open to naming help

Reply:
Heh that's fun. Umm I think this is fine then.